Add ExtensionsManager

dragssvd opened the pull request.

JakubBernaciak committed:
Add ExtensionsManager

dragssvd committed:
add DriverComposite with HashMap

dragssvd committed:
fixed per driver logger funcionality

dragssvd committed:
fix driver not showing

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/decorator/DriverDecorator.java:
```diff
@@ -13,11 +25,19 @@ public DriverDecorator(Job2dDriver driver){
     @Override
     public void setPosition(int x, int y) {
         driver.setPosition(x,y);
+        if(ExtensionsManager.isDistanceLogRunning())
+            deviceUsageManager.calculateMovingDistance(x,y);
+        if(ExtensionsManager.isLoggerRunning())
+            positionLoggingDriver.setPosition(x,y);
```
This is totally not SOLID! Consider SRP and OCP?  Does your solution at any level support it?
This code looks like a quite simple to avoid code duplication. The same type of condition, the similar methods signature. However it should not be hardcoded, but Object Oriented. 
Driver decorator is only a hint for the developers that classes implementing it are decorators and can be treated in similar way. Adding more logic to this class does not make much sense. Moreover, neither logger nor distance extensions do not need to be decorators (they can exist as child drivers within drivers composite).

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/ExtensionsManager.java:
```diff
@@ -0,0 +1,47 @@
+    public void setup(Application application) {
+        application.addComponentMenu(ExtensionsManager.class,"Extensions");
+        application.addComponentMenuElementWithCheckBox(ExtensionsManager.class, "Logger", (ActionEvent e) -> toggleLogger(), true);
+        application.addComponentMenuElementWithCheckBox(ExtensionsManager.class, "Distance log",(ActionEvent e) -> toggleDistanceLog(), true);
+
+        loggerDistanceObserver = new LoggerDistanceObserver(DeviceUsageManager.getInstance());
+        DeviceUsageManager.getInstance().addObserver(loggerDistanceObserver);
+
+    }
+
+    public void toggleLogger() {
+        DeviceUsageManager.getInstance().toggleLoggerObserver();
+        isLoggerRunning = !isLoggerRunning;
+    }
+    
+
+    public void toggleDistanceLog(){
+        isDistanceLogRunning = !isDistanceLogRunning;
+    }
```
Adding a new driver requires only calling addDriver method in DriverFeature. However, to introduce an extension developer needs to modify several classes :(. You haven't introduced a clear pattern of conduct :(.

kowallus reviewed (changes requested):
Your solution is currently bad in design. However, if you analyze it, you may find clues how to do it at least partially better.

JakubBernaciak committed:
Add new classes

We decided to create new classes instead of editing an exisiting class.
Separating functionality into new classes maintains a single responsibility for  each class.
This makes code more readable and easier to manage and understand.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/decorator/DriverDecorator.java:
```diff
@@ -25,19 +21,13 @@ public DriverDecorator(Job2dDriver driver){
     }
 
     @Override
     public void operateTo(int x, int y) {
         driver.operateTo(x,y);
-        if(ExtensionsManager.isDistanceLogRunning())
-            deviceUsageManager.calculateOperatingDistance(x,y);
-        if(ExtensionsManager.isLoggerRunning())
-            positionLoggingDriver.operateTo(x, y);
+        deviceUsageManager.calculateOperatingDistance(x,y);
```
this abstract class should not be modified and burden with device usage responsibilites

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/DeviceUsageManager.java:
```diff
@@ -28,18 +22,16 @@ private double calculateDistance(int x, int y){
-            distanceChangePublisher.notifyObservers();
-        }
+        distanceChangePublisher.notifyObservers();
+        posiotionPublisher.notifyObservers();
```
hard-coded notifications. I think that position publisher is not related to DeviceUsageManagement (and does not need to be).

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/ExtensionsManager.java:
```diff
@@ -1,47 +1,76 @@
+                manager.getDistanceChangePublisher().clearObservers();
+            }
+        }
         isDistanceLogRunning = !isDistanceLogRunning;
```
code duplication! Again, this is a code-smell, and good point for refactor.

kowallus reviewed (changes requested):
I am sorry, but I cannot see Design here. If you want I can give you a chance in October, to finish this correctly.

dragssvd committed:
Add proper addExtension

Eliminated code duplication and improved code reusability by introducing interfaces and methods.
Added 'addExtension' method to streamline extension integration.
Avoided 'instanceof' checks and direct references to specific logger/device aspects, promoting encapsulation.

kowallus reviewed (approved):
Miraculus! Congratulations :)
